File: src/com/mishiranu/dashchan/media/AnimatedImageDecoder.java
package com.mishiranu.dashchan.media;

import android.graphics.ImageDecoder;
import android.graphics.drawable.AnimatedImageDrawable;
import android.graphics.drawable.Drawable;
import java.io.File;
import java.io.IOException;

/**
 * Decodes animated GIF, PNG and WebP files via the platform {@link ImageDecoder}.
 * Throws {@link IOException} if the file is not an animated image, letting the
 * caller fall back to regular bitmap decoding.
 */
public class AnimatedImageDecoder {
	private final AnimatedImageDrawable drawable;

	public AnimatedImageDecoder(File file) throws IOException {
		Drawable drawable = ImageDecoder.decodeDrawable(ImageDecoder.createSource(file));
		if (!(drawable instanceof AnimatedImageDrawable)) {
			throw new IOException("Not an animated image");
		}
		this.drawable = (AnimatedImageDrawable) drawable;
		this.drawable.setRepeatCount(AnimatedImageDrawable.REPEAT_INFINITE);
		this.drawable.start();
	}

	public Drawable getDrawable() {
		return drawable;
	}

	public void recycle() {
		drawable.stop();
	}
}

File: src/com/mishiranu/dashchan/media/AnimatedImageDecoder.kt
package com.mishiranu.dashchan.media

import android.graphics.ImageDecoder
import android.graphics.drawable.AnimatedImageDrawable
import android.graphics.drawable.Drawable
import java.io.File
import java.io.IOException

/**
 * Decodes animated GIF, PNG and WebP files via the platform [ImageDecoder].
 * Throws [IOException] if the file is not an animated image, letting the
 * caller fall back to regular bitmap decoding.
 */
class AnimatedImageDecoder @Throws(IOException::class) constructor(file: File) {
	private val drawable: AnimatedImageDrawable

	init {
		val drawable = ImageDecoder.decodeDrawable(ImageDecoder.createSource(file))
		if (drawable !is AnimatedImageDrawable) {
			throw IOException("Not an animated image")
		}
		this.drawable = drawable
		this.drawable.repeatCount = AnimatedImageDrawable.REPEAT_INFINITE
		this.drawable.start()
	}

	fun getDrawable(): Drawable = drawable

	fun recycle() {
		drawable.stop()
	}
}
